pennylane/templates/layers.py

```diff
@@ -0,0 +1,418 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+r"""
+Layers
+======
+
+**Module name:** :mod:`pennylane.templates.layers`
+
+.. currentmodule:: pennylane.templates.layers
+
+This module contains templates for trainable `layers`. A layer is a fixed sequence of quantum gates,
+where some gates have trainable gate parameters (like rotation angles or squeezing amplitudes). This
+makes the layer `learnable` within the limits of the architecture. A variational circuit typically repeats layers.
+
+Most templates in this module have a ``Layer`` version that implements a single layer, as well as a ``Layers``
+version which calls the single layer multiple times, possibly using different hyperparameters for the
+sequence in each call.
+
+
+Qubit architectures
+-------------------
+
+Strongly entangling circuit
+***************************
+
+.. autosummary::
+
+    StronglyEntanglingLayers
+    StronglyEntanglingLayer
+
+Random circuit
+**************
+
+.. autosummary::
+
+    RandomLayers
+    RandomLayer
+
+Continuous-variable architectures
+---------------------------------
+
+Continuous-variable quantum neural network
+******************************************
+
+.. autosummary::
+
+    CVNeuralNetLayers
+    CVNeuralNetLayer
+
+Interferometer
+**************
+
+.. autosummary::
+
+    Interferometer
+
+Code details
+^^^^^^^^^^^^
+"""
+#pylint: disable-msg=too-many-branches,too-many-arguments,protected-access
+from collections.abc import Sequence
+
+from pennylane.ops import CNOT, RX, RY, RZ, Rot, Squeezing, Displacement, Kerr, Beamsplitter, Rotation
+from pennylane.qnode import QuantumFunctionError
+from pennylane.variable import Variable
+import numpy as np
+
+
+def StronglyEntanglingLayers(weights, ranges=None, imprimitive=CNOT, wires=None):
+    """A sequence of layers of type :func:`StronglyEntanglingLayer()`, as specified in :cite:`schuld2018circuit`.
+
+    The number of layers :math:`L` is determined by the first dimension of ``weights``. The template is applied to
+    the qubits specified by the sequence ``wires``.
+
+    Args:
+        weights (array[float]): array of weights of shape ``(L, len(wires), 3)``
+
+    Keyword Args:
+        ranges (Sequence[int]): sequence of ranges determining the architecture of the imprimitive gates in each layer
+        imprimitive (pennylane.ops.Operation): two-qubit gate to use, defaults to :class:`~.CNOT`
+        wires (Sequence[int]): sequence of qubit indices that the template acts on
+    """
+
+    if ranges is None:
+        ranges = [1]*len(weights)
+
+    for block_weights, block_range in zip(weights, ranges):
+        StronglyEntanglingLayer(block_weights, r=block_range, imprimitive=imprimitive, wires=wires)
+
+
+def StronglyEntanglingLayer(weights, r=1, imprimitive=CNOT, wires=None):
```

Comment:
should we mention that r=1 by default below?

Reply:
Done